crates/ra_hir/src/code_model.rs

```diff
@@ -960,6 +960,38 @@ impl ImplDef {
         db.impl_data(self.id).target_trait.clone()
     }
 
+    pub fn target_trait_substs_matches(&self, db: &dyn HirDatabase, typs: &[Type]) -> bool {
+        let type_ref = match self.target_trait(db) {
+            Some(typ_ref) => typ_ref,
+            None => return false,
+        };
+        let resolver = self.id.resolver(db.upcast());
+        let ctx = hir_ty::TyLoweringContext::new(db, &resolver);
+        let ty = Ty::from_hir(&ctx, &type_ref);
```

Comment:
this isn't quite the right approach -- what you're doing here is taking the trait reference (like `From<MyEnum>`) and resolving it as a type, so it gets interpreted as `dyn From<MyEnum>`. It kind of gives you what you want since you then use `dyn_trait_ref`, but is a bit too roundabout :slightly_smiling_face: It's also where the weirdness with the bound var below comes from. A better way is calling `db.impl_trait(self.id)`. That will give you a trait reference directly (inside a `Binders`; this thing tells you how many type variables the impl has. For example if you had `impl<T> From<T> for Foo`, the `Binders` would say that there's one variable, and then the `T` would be represented by `Ty::Bound(0)`. You can probably just ignore this for now though.)